environs/manual/provisioner.go

```diff
@@ -1,248 +1,15 @@
-		return nil, err
-	}
-	if provisioned {
-		return nil, ErrProvisioned
-	}
-
-	hc, series, err := DetectSeriesAndHardwareCharacteristics(hostname)
-	if err != nil {
-		err = fmt.Errorf("error detecting hardware characteristics: %v", err)
-		return nil, err
-	}
-
-	// There will never be a corresponding "instance" that any provider
-	// knows about. This is fine, and works well with the provisioner
-	// task. The provisioner task will happily remove any and all dead
-	// machines from state, but will ignore the associated instance ID
-	// if it isn't one that the environment provider knows about.
-
-	instanceId := instance.Id(manualInstancePrefix + hostname)
-	nonce := fmt.Sprintf("%s:%s", instanceId, uuid.String())
-	machineParams := &params.AddMachineParams{
-		Series:                  series,
-		HardwareCharacteristics: hc,
-		InstanceId:              instanceId,
-		Nonce:                   nonce,
-		Addrs:                   params.FromNetworkAddresses(addr),
-		Jobs:                    []multiwatcher.MachineJob{multiwatcher.JobHostUnits},
-	}
-	return machineParams, nil
-}
-
-var provisionMachineAgent = func(host string, icfg *instancecfg.InstanceConfig, progressWriter io.Writer) error {
-	script, err := ProvisioningScript(icfg)
-	if err != nil {
-		return err
-	}
-	return runProvisionScript(script, host, progressWriter)
-}
-
-// ProvisioningScript generates a bash script that can be
-// executed on a remote host to carry out the cloud-init
-// configuration.
-func ProvisioningScript(icfg *instancecfg.InstanceConfig) (string, error) {
-	cloudcfg, err := cloudinit.New(icfg.Series)
-	if err != nil {
-		return "", errors.Annotate(err, "error generating cloud-config")
-	}
-	cloudcfg.SetSystemUpdate(icfg.EnableOSRefreshUpdate)
-	cloudcfg.SetSystemUpgrade(icfg.EnableOSUpgrade)
-
-	udata, err := cloudconfig.NewUserdataConfig(icfg, cloudcfg)
-	if err != nil {
-		return "", errors.Annotate(err, "error generating cloud-config")
-	}
-	if err := udata.ConfigureJuju(); err != nil {
-		return "", errors.Annotate(err, "error generating cloud-config")
-	}
-
-	configScript, err := cloudcfg.RenderScript()
-	if err != nil {
-		return "", errors.Annotate(err, "error converting cloud-config to script")
-	}
-
-	var buf bytes.Buffer
-	// Always remove the cloud-init-output.log file first, if it exists.
-	fmt.Fprintf(&buf, "rm -f %s\n", utils.ShQuote(icfg.CloudInitOutputLog))
-	// If something goes wrong, dump cloud-init-output.log to stderr.
-	buf.WriteString(shell.DumpFileOnErrorScript(icfg.CloudInitOutputLog))
-	buf.WriteString(configScript)
-	return buf.String(), nil
-}
-
-func runProvisionScript(script, host string, progressWriter io.Writer) error {
-	params := sshinit.ConfigureParams{
-		Host:           "ubuntu@" + host,
-		ProgressWriter: progressWriter,
-	}
-	return sshinit.RunConfigureScript(script, params)
+// Provisioner is the process of provisioning a windows/linux machine
+type Provisioner interface {
```

Comment:
I don't think this interface is necessary for now. Please just have top-level functions in the sshprovisioner/winrmprovisioner packages. We may later want to have something stateful, but for now it's just additional boilerplate for no benefit.

It would be useful to have a type for the function though:

```go
type ProvisionMachineFunc func(ProvisionMachineArgs) (machineId string, err error)
```